File: backend/app/src/main/kotlin/net/animeta/backend/controller/v3/UpdateCategoryController.kt
package net.animeta.backend.controller.v3

import net.animeta.backend.dto.CategoryDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.User
import net.animeta.backend.repository.CategoryRepository
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.CategorySerializer
import net.animeta.backend.service.CategoryMutations
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class UpdateCategoryController(
    private val categoryRepository: CategoryRepository,
    private val categoryMutations: CategoryMutations,
    private val categorySerializer: CategorySerializer
) {
    data class Params(
        val id: Int,
        val name: String?
    )
    data class Result(
        val category: CategoryDTO
    )

    @PostMapping("/v3/UpdateCategory")
    @Transactional
    fun handle(@RequestBody params: Params, @CurrentUser currentUser: User): Result {
        val category = categoryRepository.findById(params.id).orElseThrow { ApiException.notFound() }
        if (currentUser.id != category.user.id) {
            throw ApiException.permissionDenied()
        }
        if (params.name != null) {
            categoryMutations.updateName(category, params.name)
        }
        return Result(categorySerializer.serialize(category))
    }
}

File: backend-ts/src/controllers/update_category.controller.ts
import { Body, Controller, HttpCode, HttpStatus, Post } from "@nestjs/common";
import { InjectEntityManager } from "@nestjs/typeorm";
import { CategoryDTO } from "shared/types_generated";
import { CurrentUser } from "src/auth/decorators";
import { Category } from "src/entities/category.entity";
import { User } from "src/entities/user.entity";
import { CategorySerializer } from "src/serializers/category.serializer";
import { EntityManager } from "typeorm";
import { ApiException } from "./exceptions";

type Params = {
  id: number;
  name?: string | null;
}

type Result = {
  category: CategoryDTO;
}

@Controller('/api/v4/UpdateCategory')
export class UpdateCategoryController {
  constructor(
    @InjectEntityManager() private em: EntityManager,
    private categorySerializer: CategorySerializer,
  ) {}

  @Post()
  @HttpCode(200)
  async handle(
    @Body() params: Params,
    @CurrentUser() currentUser: User,
  ): Promise<Result> {
    const category = await this.em.findOne(Category, params.id)
    if (!category)
      throw ApiException.notFound()
    if (currentUser.id !== category.user_id)
      throw ApiException.permissionDenied()
    if (params.name != null) {
      if (params.name.trim() === '')
        throw new ApiException("분류 이름을 입력하세요.", HttpStatus.BAD_REQUEST)
      category.name = params.name
    }
    await this.em.save(category)
    return {category: this.categorySerializer.serialize(category)}
  }
}
